dartagnan/src/main/java/com/dat3m/dartagnan/expression/BNonDet.java

```diff
@@ -23,25 +27,30 @@ public IConst reduce() {
 	}
 
     @Override
-    public Expr toZ3Int(Event e, Context ctx) {
-    	Expr e1 = precision > 0 ? ctx.mkBV(1, precision) : ctx.mkInt(1);
-    	Expr e2 = precision > 0 ? ctx.mkBV(0, precision) : ctx.mkInt(0);
-        return ctx.mkITE(toZ3Bool(e, ctx), e1, e2);
+    public Formula toIntFormula(Event e, SolverContext ctx) {
+    	FormulaManager fmgr = ctx.getFormulaManager();
+		Formula e1 = precision > 0 ? 
+    			fmgr.getBitvectorFormulaManager().makeBitvector(precision, BigInteger.ONE) : 
+    			fmgr.getIntegerFormulaManager().makeNumber(BigInteger.ONE);
+    	Formula e2 = precision > 0 ?
+    			fmgr.getBitvectorFormulaManager().makeBitvector(precision, BigInteger.ZERO) : 
+    			fmgr.getIntegerFormulaManager().makeNumber(BigInteger.ZERO);
+        return fmgr.getBooleanFormulaManager().ifThenElse(toBoolFormula(e, ctx), e1, e2);
```

Comment:
It would be clearer to do the "precision > 0" check only once and have two branches that initialize e1 and e2 to either Integers or BitVectors. This better shows that either both expressions will be Int or both will be BV.
```
Formula e1, e2;
if (precision > 0) {
    e1 = BV(1);
    e2 = BV(0);
} else {
    e1 = Int(1);
    e2 = Int(0);
}
return ifThenElse(toBoolFormula(e, ctx), e1, e2);
```

Reply:
About to be pushed